data/sea/30-array.h

```diff
@@ -203,15 +203,16 @@ static ARRAY_T(t) INLINE ARRAY_FUN(t,put_immutable)
         ARRAY_T(t) arr = (ARRAY_T(t))imempool_alloc(pool, bytes_new);           \
         memcpy(arr, x, bytes_old);                                              \
         x = arr;                                                                \
+        x->count = ix + 1;                                                      \
                                                                                 \
     } else if (ix >= count) {                                                   \
-      ARRAY_T(t) arr = (ARRAY_T(t))imempool_alloc(pool, bytes_old);             \
-      memcpy(arr, x, bytes_old);                                                \
-      x = arr;                                                                  \
+        ARRAY_T(t) arr = (ARRAY_T(t))imempool_alloc(pool, bytes_old);           \
+        memcpy(arr, x, bytes_old);                                              \
+        x = arr;                                                                \
+        x->count = ix + 1;                                                      \
                                                                                 \
     }                                                                           \
                                                                                 \
-    x->count = ix + 1;                                                          \
     ARRAY_PAYLOAD(t,x)[ix] = v;                                                 \
```

Comment:
ah, yeah, so there are two versions now: put mutable and put immutable. put immutable is very rarely used, and most cases are optimised to mutable ones. the immutable versions were added because we now expose map and array update from the source language, so some of those uses end up being immutable.
we probably haven't actually encountered any cases where it would end up using this though..